macaroonbakery/codec.py

```diff
@@ -65,22 +70,91 @@ def _encode_caveat_v1(condition, root_key, third_party_pub_key, key):
             Base64Encoder).decode('ascii'),
         'Nonce': base64.b64encode(nonce).decode('ascii'),
         'Id': base64.b64encode(encrypted).decode('ascii')
-    }))).decode('ascii')
+    })))
+
+
+PUBLIC_KEY_PREFIX_LEN = 4
+KEY_LEN = 32
+
+
+def _encode_caveat_v2_v3(version, condition, root_key, third_party_pub_key,
+                         key, ns):
+    '''Create a version 2 or version 3 third-party caveat.
+
+    The format has the following packed binary fields (note
+    that all fields up to and including the nonce are the same
+    as the v2 format):
+
+        version 2 or 3 [1 byte]
+        first 4 bytes of third-party Curve25519 public key [4 bytes]
+        first-party Curve25519 public key [32 bytes]
+        nonce [24 bytes]
+        encrypted secret part [rest of message]
+
+    The encrypted part encrypts the following fields
+    with box.Seal:
+
+        version 2 or 3 [1 byte]
+        length of root key [n: uvarint]
+        root key [n bytes]
+        length of encoded namespace [n: uvarint] (Version 3 only)
+        encoded namespace [n bytes] (Version 3 only)
+        condition [rest of encrypted part]
+    '''
+    ns_data = b''
+    if version >= bakery.BAKERY_V3:
+        ns_data = ns.serialize()
+    data = six.int2byte(version)
```

Comment:
I think this code would probably be quite a bit more efficient if you started with a bytearray
and converted it to bytes at the end, rather than copying the data bytes each time.